### PR TITLE
Move SignalGenerator from Hopper example to osimCommon.

### PR DESCRIPTION
**Pull request by chrisdembia**
This PR addresses #1500: The SignalGenerator class used in the hopper example is generally useful, so I have moved it the osimCommon library. I also added tests for this class.

Ideally, the SignalGenerator could support types other than `double`, but Function can only provide `double`s.

### CODE REVIEW

File: OpenSim/Common/Test/testComponentInterface.cpp [hunk 1]
```diff
@@ -1926,6 +1928,49 @@ void testAliasesAndLabels() {
...
+    const double phase = 0.3;
+    signalGen->set_function(Sine(amplitude, omega, phase));
+    world.addComponent(signalGen);
+    
+    auto* reporter = new TableReporter();
+    reporter->addToReport(signalGen->getOutput("signal"));
+    world.addComponent(reporter);
+
+    // Build the system.
+    // Check that the value of the input is the same as before.
```
Comment (tkuchida) [inline OpenSim/Common/Test/testComponentInterface.cpp]: There's no check (e.g., `SimTK_TEST_EQ()`) in this block and the system is being realized to only `Stage::Model` so it isn't clear (to me, at least) how L1951–1954 contribute to checking the value of the input. Perhaps remove this comment?

File: OpenSim/Common/Test/testComponentInterface.cpp [hunk 2]
```diff
@@ -1926,6 +1928,49 @@ void testAliasesAndLabels() {
...
+    MultibodySystem system;
+    world.buildUpSystem(system);
+    State s = system.realizeTopology();
+    system.realize(s, Stage::Model);
+
+    // "Simulate."
+    const int numTimeSteps = 5;
+    for (int i = 0; i < numTimeSteps; ++i) {
```
Comment (tkuchida) [inline OpenSim/Common/Test/testComponentInterface.cpp]: For `numTimeSteps` to mean "the number of time steps taken", I suggest changing to `<=` (with corresponding changes to L1965–1966).

File: OpenSim/Common/Test/testSignalGenerator.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+void testSignalGenerator() {
+}
```
Comment (tkuchida) [inline OpenSim/Common/Test/testSignalGenerator.cpp]: Perhaps include a comment explaining why this method is empty?

Comment (klshrinidhi) [inline OpenSim/Common/Test/testSignalGenerator.cpp]: Wow!! Markdown is awesome!!

File: OpenSim/Common/Test/testComponentInterface.cpp (hunk 1 above)
Comment (chrisdembia) [inline OpenSim/Common/Test/testComponentInterface.cpp]: Removed the comment...C&P error.

File: OpenSim/Common/Test/testComponentInterface.cpp (hunk 2 above)
Comment (chrisdembia) [inline OpenSim/Common/Test/testComponentInterface.cpp]: numTimeSteps -> numTimePoint.

File: OpenSim/Common/Test/testSignalGenerator.cpp (hunk 3 above)
Comment (chrisdembia) [inline OpenSim/Common/Test/testSignalGenerator.cpp]: Didn't mean to commit this file!

**Comment by chrisdembia** [comment]
@tkuchida I have tried to address your concerns 🏉

**Comment by tkuchida** [comment]
> I have tried to address your concerns

Looks good. Super pull (request) :trophy:

File: OpenSim/Common/Test/testComponentInterface.cpp [hunk 4]
```diff
@@ -1926,6 +1928,48 @@ void testAliasesAndLabels() {
     SimTK_TEST(foo->getInput("listInput1").getLabel(1) == "thud");
 }
 
+void testSignalGenerator() {
```
Comment (aseth1) [inline OpenSim/Common/Test/testComponentInterface.cpp]: :+1: for the test case. It really doesn't belong in `testComponentInterface`. We could start a new test suite (maybe `testFunctions`) for `Function`s and some of its primary clients, e.g. `SignalGenerator` to verify their correctness.

File: OpenSim/Tests/Components/testComponents.cpp [hunk 5]
```diff
@@ -122,6 +122,8 @@ int main()
     }
     */
 
+    availableComponents.push_back(new SignalGenerator());
```
Comment (aseth1) [inline OpenSim/Tests/Components/testComponents.cpp]: :+1:

**Comment by chrisdembia** [comment]
Thanks @aseth1. I agree that there should be a separate test case, and I think `testFunctions` is a good idea. 

I want to have `TheWorld` accessible in that separate test case (this is why I had sloppily put the test in testComponentInterface). I will either make a separate "World" class to be used in other test cases in the Common library, or I will copy `TheWorld` from `testComponentInterface`. I'll try to see which option makes more sense.

**Comment by chrisdembia** [comment]
@aseth1 This is ready for review again.

File: OpenSim/Common/Test/testFunctions.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#include <OpenSim/Common/SignalGenerator.h>
+#include <OpenSim/Common/Reporter.h>
+
+#include "ComponentsForTesting.h"
+
+using namespace OpenSim;
+using namespace SimTK;
+
+
+void testSignalGenerator() {
```
Comment (aseth1) [inline OpenSim/Common/Test/testFunctions.cpp]: I think this is fine for testing the SignalGenerator. 

I was also thinking that we could have:
```
testFunction(const Function& func, const std::vector<double>& times, 
                 const std::vector<double>& expectedValues) 
{
    //compare function evaluated at times provided to expected values provided
}

Comment (aseth1) [inline OpenSim/Common/Test/testFunctions.cpp]: Probably best to do consider as a subsequent PR.

Comment (aseth1) [inline OpenSim/Common/Test/testFunctions.cpp]: I added an issue #1518 to address Functions. I am merging this PR.

**Comment by chrisdembia** [comment]
Thank you @aseth1 .

**Pull request merged**